src/main/scala/shine/OpenCL/HostCodeGenerator.scala

```diff
@@ -0,0 +1,200 @@
+package shine.OpenCL
+
+import arithexpr.arithmetic
+import shine.C
+import shine.C.AST._
+import shine.DPIA.Phrases._
+import shine.DPIA.Types._
+import shine.DPIA.VarType
+import shine.OpenCL.compilation.HostManagedBuffers
+import shine.OpenCL.primitives.imperative.{KernelCallCmd, NewManagedBuffer}
+
+import scala.collection.{immutable, mutable}
+
+object HostCodeGenerator {
+  def apply(): HostCodeGenerator =
+    new HostCodeGenerator(mutable.ListBuffer[Decl](), immutable.Map[String, arithmetic.Range]())
+}
+
+case class HostCodeGenerator(override val decls: C.CodeGenerator.Declarations,
+                             override val ranges: C.CodeGenerator.Ranges)
+  extends C.CodeGenerator(decls, ranges)
+{
+  override def name: String = "OpenCL Host"
+
+  override def cmd(env: Environment): Phrase[CommType] => Stmt = {
+    case KernelCallCmd(name, LocalSize(ls), GlobalSize(gs), output, args) =>
+      val arg_count = 1 + args.size
+      output |> acc(env, Nil, outputC =>
+        expSeq(args, env, argsC =>
+          C.AST.Block(
+            Seq(
```

Comment:
Some comments explaining the individal code lines/blocks that get generated would be helpful.

Reply:
Will do.

Reply:
I started by using more temporary values whose name might be enough to clarify the intent, I can also add comments if it remains hard to read.

Reply:
:+1: I think that's fine.